### PR TITLE
 Refactor presto/prestoadmin installation out of base_product_case.

### PR DESCRIPTION
**Pull request by ebd2**
In order to support writing product tests for the YARN integration with
slider, we need to refactor the product tests so that we can reuse the
existing docker/configurable cluster logic with slider-based presto
installation and configuration. This change extracts the installation of
presto-admin, presto, and setting the default topology from base_product_case
and moves it to individual modules.

A quick guide to the changes:
In the beginning base_product_case contained the logic for installing presto and presto-admin on a cluster, and for setting the topology. This worked, and code in base_product_case handled setting up clusters with and without presto for various tests.
With the upcoming changes to support YARN integration, we need to use base_product_test as a base class for tests that install presto using slider. Those tests will share the need to set up a cluster and install presto-admin, but the configuration and presto installation will be different. To make it a little easier to do that work I've done the following:
Pulled most of the installation logic out of base_product_case and separated it into modules that install the following:
* prestoadmin_installer: install presto-admin
* topology_installer: install the topology
* presto_installer: install presto

A cluster type is now defined by the set of installers that have to be run against a bare cluster to bring the cluster to the requested state. Each of docker_cluster and configurable_cluster can supply a bare cluster to base_product_case, which is then responsible for running the installers before returning the cluster to the caller.
Additionally, docker_cluster knows how to commit a cluster as an image to use to create clusters again in the future without going through the expensive installation process again. Some of this capability existed before, but we were only using it for a fully-installed presto cluster. In particular, we were recreating the base image every time we created a base cluster. The new behavior is to save an image for reuse for every cluster type, including for the bare cluster.

The tests currently run against a docker cluster. I'll update the PR with any changes needed to run against a configurable_cluster later.

Update:
---

I have the smoke tests running against a vagrant cluster and so far things appear to be OK. I'm cautiously optimistic that there won't be a big change to get configurable cluster fully working.

Also, can I please get a look from @cawallin, @rschlussel.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -45,7 +45,7 @@ clean-test:
 	rm -f .coverage
 	rm -fr htmlcov/
 	rm -fr tmp
-	docker rmi -f teradatalabs/centos-presto-test-master teradatalabs/centos-presto-test-slave || true
+	for image in $(docker images | awk '/teradatalabs\/pa_test/ {print $3}'); do docker rmi -f $image ; done
```
Comment (cawallin) [inline Makefile]: nice!

**Comment by anusudarsan** [comment]
lgtm mod comments. Are you planning to update PR #69 after this?

File: tests/docker_cluster.py [hunk 2]
```diff
@@ -334,67 +336,80 @@ def exec_cmd_on_host(self, host, cmd, raise_error=True, tty=False):
...
-            return presto_cluster
+
+        centos_cluster.start_containers(master_name, slave_name)
+
+        return centos_cluster
 
     @staticmethod
-    def start_base_cluster():
+    def start_existing_images(cluster_type):
+        DC = DockerCluster
```
Comment (cawallin) [inline tests/docker_cluster.py]: should be lower case. I'm not sure I like this abbreviation; it doesn't fit in with the presto style.

File: tests/docker_cluster.py [hunk 3]
```diff
@@ -334,67 +336,80 @@ def exec_cmd_on_host(self, host, cmd, raise_error=True, tty=False):
...
-            'teradatalabs/centos6-ssh-test',
-            'teradatalabs/centos6-ssh-test'
-        )
+
+        centos_cluster.start_containers(master_name, slave_name)
         return centos_cluster
 
     @staticmethod
-    def _check_for_presto_images():
+    def _check_for_images(master_image_name, slave_image_name):
```
Comment (cawallin) [inline tests/docker_cluster.py]: In the future, can you separate out the rename changes from the substantial changes? it makes it easier to review

File: tests/docker_cluster.py [hunk 4]
```diff
@@ -334,67 +336,80 @@ def exec_cmd_on_host(self, host, cmd, raise_error=True, tty=False):
...
+            if master_image_name in image['RepoTags'][0]:
                 has_master_image = True
-            if INSTALLED_PRESTO_TEST_SLAVE_IMAGE in image['RepoTags'][0]:
+            if slave_image_name in image['RepoTags'][0]:
                 has_slave_image = True
         return has_master_image and has_slave_image
 
+    def commit_images(self, cluster_type):
+        self.client.commit(self.master,
+                           self._get_master_image_name(cluster_type))
```
Comment (cawallin) [inline tests/docker_cluster.py]: make sure to run lint, this is a bit suspicious indentation (though it could just be github)

File: tests/docker_cluster.py [hunk 5]
```diff
@@ -419,14 +434,6 @@ def write_content_to_host(self, content, path, host):
     def copy_to_host(self, source_path, dest_host):
         shutil.copy(source_path, self.get_local_mount_dir(dest_host))
 
-    @nottest
-    def clean_up_presto_test_images(self):
```
Comment (cawallin) [inline tests/docker_cluster.py]: I think it still may be necessary to clean up the test images -- that way if you're running product tests from intellij it deletes the old images.

File: tests/product/base_product_case.py [hunk 6]
```diff
@@ -108,69 +106,55 @@ class BaseProductTestCase(BaseTestCase):
...
         self.cluster = None
+        self.default_keywords = {}
+
+    def _run_installers(self, installers):
+        cluster = self.cluster
+        for installer in installers:
+            dependencies = installer.get_dependencies()
+
+            for dependency in dependencies:
+                dependency.assert_installed(self, cluster.master)
```
Comment (cawallin) [inline tests/product/base_product_case.py]: why does this assume cluster.master?

File: tests/product/base_product_case.py [hunk 7]
```diff
@@ -17,34 +17,32 @@
...
 PRESTO_VERSION = r'presto-main:.*'
 RETRY_TIMEOUT = 120
 RETRY_INTERVAL = 5
 
-DUMMY_RPM_NAME = 'dummy-rpm.rpm'
-
 
 class BaseProductTestCase(BaseTestCase):
+    _cluster_types = {
+        'bare': [],
```
Comment (cawallin) [inline tests/product/base_product_case.py]: can you extract these as constants? e.g. 'bare', 'pa_only', etc

File: tests/product/presto_installer.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+            cluster = testcase.cluster
+
+        try:
+            check_rpm = cluster.exec_cmd_on_host(
+                container, 'rpm -q presto-server-rpm')
+            testcase.assertRegexpMatches(
+                check_rpm, RPM_BASENAME + '\n', msg=msg
+            )
+        except OSError as e:
+            if isinstance(testcase.cluster, DockerCluster):
```
Comment (cawallin) [inline tests/product/presto_installer.py]: we can probably remove this commit, @mattsfuller probably found the underlying cause of presto-admin failures

File: tests/product/test_control.py [hunk 9]
```diff
@@ -142,21 +141,23 @@ def test_server_restart_various_states(self):
             running_host=self.cluster.internal_slaves[0])
 
     def test_start_stop_restart_coordinator_down(self):
-        self.setup_cluster()
-        self.install_presto_admin(self.cluster)
+        installer = PrestoInstaller(self)
```
Comment (cawallin) [inline tests/product/test_control.py]: small, but could you rename this to presto_installer so it's more clear?

File: tests/product/test_server_uninstall.py [hunk 10]
```diff
@@ -1,4 +1,4 @@
-# -*- coding: utf-8 -*-
```
Comment (cawallin) [inline tests/product/test_server_uninstall.py]: why did you remove this? check if others were also removed

**Comment by cawallin** [comment]
looks good % small comments and running lint

File: Makefile [hunk 11]
```diff
@@ -45,7 +45,7 @@ clean-test:
 	rm -f .coverage
 	rm -fr htmlcov/
 	rm -fr tmp
-	docker rmi -f teradatalabs/centos-presto-test-master teradatalabs/centos-presto-test-slave || true
+	for image in $$(docker images | awk '/teradatalabs\/pa_test/ {print $$3}'); do docker rmi -f $$image ; done
 	echo "Note: The above command is just cleaning up a Docker image that may not exist. If the command fails, it is not a problem."
```
Comment (rschlussel) [inline Makefile]: We can get rid of this warning now, right?  Because it's only iterating through images that exist.

File: tests/docker_cluster.py (hunk 4 above)
Comment (ebd2) [inline tests/docker_cluster.py]: lint passes, I think this one is github

File: tests/product/base_product_case.py (hunk 6 above)
Comment (ebd2) [inline tests/product/base_product_case.py]: How about if cluster exposes a get_master() method for installers and then they can pick?

File: tests/product/test_server_uninstall.py (hunk 10 above)
Comment (ebd2) [inline tests/product/test_server_uninstall.py]: Spurious change. I'll revert it.

File: tests/docker_cluster.py (hunk 5 above)
Comment (ebd2) [inline tests/docker_cluster.py]: Let's sync on this. It's not clear to me how the place we're calling this currently is accomplishing cleaning up the images when we run this from intellij

File: Makefile (hunk 11 above)
Comment (ebd2) [inline Makefile]: I've updated the warning to reflect how it might fail now and what to do if it does.

**Comment by cawallin** [comment]
looks good, please squash the feedback commits together
